# COMP90016 Assignment2 - Variant Calling and Phasing

## Setup

This project requires Python >= 2.7 < 3.0.

## Running

The projects bin directory holds 4 command line tools as listed below

Each command line tool is self documenting by supplying the -h option as below

```
bin/snv_heterozygote.py -h
```

### Task 1

The command line tools for Task 1 are bin/snv_heterozygote.py and bin/snv_outfile_stats.py

```
 bin/snv_heterozygote.py <BAM_FILE> --quality_filter <QUALITY_CUTOFF>
```

Task 1 requires running the heterozygote caller with quality filter set to 0, which mimics the bahavior of the week5 lab, and 20. 

Once run on the samples, the output can be given to bin/snv_outfile_stats.py to generate some stats and a graph

```
 bin/snv_outfile_stats.py <OUTFILE_FROM_SNV_HETEROZYGOTE>
```

### Task 2

Task 2 builds on Task 1 by utilising the reference in FASTA format along with the sample BAM. It also outputs to a vcf format

```
bin/snv_reference.py <BAM_FILE> --reference_file <REFERENCE_FILE>
```

### Task 3

The Task 3 phaser requires a VCF file generated by Task2 as well as the original BAM file

```
bin/snv_reference.py <BAM_FILE> --vcf <TASK2_VCF_FILE>
```


## Running the Tests

Some tests (although not exhaustive ) have been written and are supplied in the tests directory. These can be run by the following command

```
  python -m unittest discover tests
```


